File: suyeonsu/Hash_전화번호목록.py
# [방법1] 정렬 -> 효율성 테스트 결과가 [방법2]보다 좋았음
def solution(phone_book):
    phone_book.sort()
    for a, b in zip(phone_book, phone_book[1:]):
        if b.startswith(a): return False
    return True

  
# [방법2] Hash
def solution(phone_book):
    h = {num:1 for num in phone_book}
    for num in phone_book:
        pre = ''
        for k in num:
            pre += k
            if pre in h and pre != num:
                return False
    return True
  
  
"""
문자열 리스트 정렬시 사전순으로 정렬됨
startswith(시작하는 문자(열), 시작지점)
endswith(끝나는 문자(열), 문자열의 시작, 문자열의 끝)
"""

File: suyeonsu/Hash_전화번호목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        boolean answer = true;
        Arrays.sort(phone_book);
        for(int i=0; i<phone_book.length-1; i++) {
            if(phone_book[i+1].length() >= phone_book[i].length() && 
                    phone_book[i+1].substring(0, phone_book[i].length()).equals(phone_book[i])) {
                answer = false;
                break;
            }
        }
        return answer;
    }
}
